### PR TITLE
Problem: can't get back tx result of false-failed tx

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #437
    Solution:
    - Support fetch tx receipts by replaying block

> wait for https://github.com/crypto-org-chain/cronos/pull/432 to get merged first.

New API:

```
$ curl -X POST --data '{"jsonrpc":"2.0","method":"cronos_replayBlock","params":["0x18"],"id":1}' -H "Content-Type: application/json" $ETH_RPC_URL
```

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: x/cronos/keeper/grpc_query.go [hunk 1]
```diff
@@ -40,3 +42,68 @@ func (k Keeper) DenomByContract(goCtx context.Context, req *types.DenomByContrac
...
+
+// ReplayBlock replay the eth messages in the block to recover the results of false-failed txs.
+func (k Keeper) ReplayBlock(goCtx context.Context, req *types.ReplayBlockRequest) (*types.ReplayBlockResponse, error) {
+	rsps := make([]*evmtypes.MsgEthereumTxResponse, 0, len(req.Msgs))
+
+	// prepare the block context, the multistore version should be setup already in grpc query context.
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	ctx = ctx.WithBlockHeight(req.BlockNumber)
+	ctx = ctx.WithBlockTime(req.BlockTime)
+	ctx = ctx.WithHeaderHash(common.Hex2Bytes(req.BlockHash))
```
Comment (JayT106) [inline x/cronos/keeper/grpc_query.go:54]: maybe do 
`ctx := sdk.UnwrapSDKContext(goCtx).WithBlockHeight(req.BlockNumber).WithBlockTime(req.BlockTime).WithHeaderHash(common.Hex2Bytes(req.BlockHash))`
pattern?

Comment (yihuang) [reply x/cronos/keeper/grpc_query.go]: sounds good, changed.

File: x/cronos/keeper/grpc_query.go [hunk 2]
```diff
@@ -40,3 +42,68 @@ func (k Keeper) DenomByContract(goCtx context.Context, req *types.DenomByContrac
...
+			evmDenom,
+			homestead,
+			istanbul,
+			london,
+		); err != nil {
+			return nil, err
+		}
+
+		// increase nonce
+		acc := k.accountKeeper.GetAccount(ctx, msg.GetFrom())
```
Comment (JayT106) [inline x/cronos/keeper/grpc_query.go:94]: I think there should verify the acc not equal to nil

File: x/cronos/rpc/api.go [hunk 3]
```diff
@@ -68,51 +74,61 @@ func NewCronosAPI(
...
-	api.logger.Debug("cronos_getTransactionReceiptsByBlock", "blockNrOrHash", blockNrOrHash)
-
-	blockNum, err := api.getBlockNumber(blockNrOrHash)
+func (api *CronosAPI) GetBlockDetail(blockNrOrHash rpctypes.BlockNumberOrHash) (
+	resBlock *coretypes.ResultBlock,
+	blockNumber int64,
+	blockHash string,
+	blockRes *coretypes.ResultBlockResults,
+	baseFee *big.Int,
+	err error,
```
Comment (JayT106) [inline x/cronos/rpc/api.go:93]: why we are returning  `blockNumber ` and `blockHash `, the `resBlock` already contains these data.
Or should we create a new struct to return the data the client requires instead return everything?

Comment (JayT106) [reply x/cronos/rpc/api.go:93]: If this API is only called by `GetTransactionReceiptsByBlock `, should we not expose this API to the public?

Comment (yihuang) [reply x/cronos/rpc/api.go:93]: done, changed to private.

File: x/cronos/keeper/grpc_query.go (hunk 2 above)
Comment (yihuang) [reply x/cronos/keeper/grpc_query.go:94]: done, added a `nil` check.

**Review by JayT106 — APPROVED** [review]
LGTM, maybe @leejw51crypto @thomas-nguy can have a look.

**Pull request merged**